examples/hybrid_jobs/6_QNSPSA_optimizer_with_embedded_simulator/qnspsa_with_embedded_simulator.ipynb

```diff
@@ -0,0 +1,1096 @@
+     ]
+    }
+   ],
+   "source": [
+    "!sed -n 106,119p source_scripts/QNSPSA.py"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "da05eaba",
+   "metadata": {},
+   "source": [
+    "To confirm the optimizer works, let's work with a toy QAOA maximum cut example [6]. The maximum cut problem is to partition the nodes of a graph into two subsets such that the number of edges going between the subsets is maximized. The QAOA max cut problem can be conveniently set up with the [Pennylane's QAOA module](https://pennylane.readthedocs.io/en/stable/code/api/pennylane.qaoa.cost.maxcut.html). You can find a more detailed and visualized example [here](https://github.com/aws/amazon-braket-examples/blob/17ba255396f4c2bf773dde99a5c46486e822776e/examples/hybrid_jobs/2_Using_PennyLane_with_Braket_Jobs/Using_PennyLane_with_Braket_Jobs.ipynb)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "id": "44ddf074",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# QAOA max cut problem set up.\n",
+    "nodes = n_qubits = 4\n",
+    "edges = 4\n",
+    "seed = 1\n",
+    "depth = 2\n",
+    "\n",
+    "g = nx.gnm_random_graph(nodes, edges, seed=seed)\n",
+    "cost_h, mixer_h = qaoa.maxcut(g)\n",
+    "\n",
+    "# Defining device to be the pennylane lightning local simulator\n",
+    "dev = qml.device(\"lightning.qubit\", wires=n_qubits)\n",
+    "\n",
+    "def qaoa_layer(gamma, alpha):\n",
+    "    qaoa.cost_layer(gamma, cost_h)\n",
+    "    qaoa.mixer_layer(alpha, mixer_h)\n",
+    "    \n",
+    "def qaoa_circuit(params, n_qubits, depth):\n",
+    "    #initalizing all qubits into +X eigenstate. \n",
+    "    for w in range(n_qubits):\n",
+    "        qml.Hadamard(wires=w)\n",
+    "    gammas = params[0]\n",
+    "    alphas = params[1]\n",
+    "    #stacking building blocks for depth times.\n",
+    "    qml.layer(qaoa_layer, depth, gammas, alphas)\n",
+    "    \n",
+    "# Defining ansatz and cost function\n",
+    "@qml.qnode(dev)\n",
+    "def cost_function(params):\n",
+    "    qaoa_circuit(params, n_qubits, depth)\n",
+    "    return qml.expval(cost_h) "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "4e3588c7",
+   "metadata": {},
+   "source": [
+    "In this QAOA problem, the loss is given by the expectation value of the cost Hamiltonian. The opposite of the loss (rounded to integer) represents the number of edges between the two subsets from a given cut, while the two subsets are specified by the $Z$ values of each qubit (node). Nodes with the same $Z$ value are of the same subset. Finding the maximum cut is then equivalent to finding the mininum-energy state of the cost Hamiltonian, and this is done with a hybrid algorithm with the QAOA ansatz. The building block of the QAOA ansatz is a cost layer followed with a mixer layer. This building block repeats for $p$ times. A larger $p$ value leads to better approximation, but also more costly experiments."
```

Comment:
```suggestion
    "In QAOA, the loss is given by the expectation value of the cost Hamiltonian. The opposite of the loss (rounded to integer) represents the number of edges between the two subsets from a given cut, while the two subsets are specified by the $Z$ values of each qubit (node). Nodes with the same $Z$ value are of the same subset. Finding the maximum cut is then equivalent to finding the mininum-energy state of the cost Hamiltonian, which is done variationally in the QAOA algorithm. The basic building block of the QAOA ansatz is a cost layer followed with a mixer layer. This building block repeats for $p$ times resulting in the final ansatz circuit. A larger $p$ value leads generally to better approximations of the solution in the ideal case, however, the noise on real QPUs will prevent you from going to higher values since the circuits get deeper with increasing $p$. In practice, a lot of experimentation is required to find the best values for your problem and device choice."
```

Reply:
Do you agree with my edits here?